Added dot feature to legend as an alternative to the default rectangles

cbakkeby opened the pull request.

cbakkeby commented:
In order to better accommodate for dot map legends, I added an if/else block (removing the switch() that was there previously) to allow for creating legend 'points' (circles of variable size/color) instead of just the default rectangles. Please see the attached screenshot for an example of a BPG dotmap / heatmap combo with a legend created using this updated legend.grob.R

![image](https://user-images.githubusercontent.com/44226915/121475032-c481d180-c979-11eb-986a-a77d8172dc34.png)

pboutros commented on R/legend.grob.R:
```diff
@@ -80,154 +80,163 @@ legend.grob <- function(
+					col = colorRamp,
+					width = if (!is.null(legendi[['angle']]) && legendi[['angle']] != 0) { legendi[['height']] } else { legendi[['width']] },
+					labels = list(
+						labels = if (is.null(legendi[['labels']])) { c('') } else { legendi[['labels']] },
+						at = if (is.null(legendi[['at']])) { NULL } else { legendi[['at']] },
+						cex = if (is.null(legendi[['cex']])) { 0.8 } else { legendi[['cex']] },
+						rot =  if (is.null(legendi[['labels.rot']])) { 0 } else { legendi[['labels.rot']] }
+						)
+					);
+
+
+				color.key.grob <- draw.colorkey(
+						key = legend.key,
+						draw = FALSE
+						);
```
check indentation, should be four spaces per indent level

pboutros commented on R/legend.grob.R:
```diff
@@ -80,154 +80,163 @@ legend.grob <- function(
-							);
-						}
+			# Figure out where this legend should go in the layout (ordered row-wise)
+			legend.row <- if (i %% layout[1] == 0) { i / layout[1] } else { floor(i / layout[1]) + 1 }
+			legend.col <- if (i %% layout[1] == 0) { layout[1] } else { i %% layout[1] }
```
This needs explanation, why are both rows layout[1] rather than one being layout[0] and layout[1] respectively?

pboutros commented on R/legend.grob.R:
```diff
@@ -80,154 +80,163 @@ legend.grob <- function(
+				# Draw a grob representing the title
+				title.x.coord <- 0.5;
+
+				if (title.just == 'left') {
```
constants go on the left of comparison operations

pboutros commented on R/legend.grob.R:
```diff
@@ -80,154 +80,163 @@ legend.grob <- function(
+						cex = title.cex,
+						fontface = title.fontface,
+						fontfamily = font.family,
+						fill = 'black'
```
why is fill required to be black and not parameterized?

pboutros commented on R/legend.grob.R:
```diff
@@ -80,154 +80,163 @@ legend.grob <- function(
+					unitTo = 'lines',
+					valueOnly = TRUE
+					);
+
+
```
extra newline

pboutros commented on R/legend.grob.R:
```diff
@@ -80,154 +80,163 @@ legend.grob <- function(
+				legend.grob.final <- packGrob(
+					frame = legend.grob.final,
+					grob = title.grob,
+					row = 3 * (legend.row - 1) + 1,
+					col = 2 * (legend.col - 1) + 1,
```
needs comments explaining why 3* and 2* on lines 125 and 126 respectively

dan-knight replied:
This behaviour comes from the [initial commit](https://github.com/uclahs-cds/public-R-BoutrosLab-plotting-general/commit/d31b7aa393bc8d0712bbec8d7a77dffe4e703f67) (moving the code from SVN).

pboutros commented on R/legend.grob.R:
```diff
@@ -80,154 +80,163 @@ legend.grob <- function(
+					row = 3 * (legend.row - 1) + 1,
+					col = 2 * (legend.col - 1) + 1,
+					height = max(
+						legend.grob.final$framevp$layout$heights[3 * (legend.row - 1) + 1],
```
ditto, needs comment explaining why 3*

pboutros commented on R/legend.grob.R:
```diff
@@ -80,154 +80,163 @@ legend.grob <- function(
+			# the second column is the corresponding text labels
+			if (!is.null(legendi[['continuous']]) && legendi[['continuous']] == TRUE) {
+
+				legendi[['height']] <- if (is.null(legendi[['height']])) { 2 } else { legendi[['height']] };
+				legendi[['width']] <- if (is.null(legendi[['width']])) { 2 } else { legendi[['width']] };
```
alighn things like this vertically with padding spaces (e.g. add one space before the <- on line 143

pboutros commented on R/legend.grob.R:
```diff
@@ -80,154 +80,163 @@ legend.grob <- function(
+					col = colorRamp,
+					width = if (!is.null(legendi[['angle']]) && legendi[['angle']] != 0) { legendi[['height']] } else { legendi[['width']] },
+					labels = list(
+						labels = if (is.null(legendi[['labels']])) { c('') } else { legendi[['labels']] },
+						at = if (is.null(legendi[['at']])) { NULL } else { legendi[['at']] },
+						cex = if (is.null(legendi[['cex']])) { 0.8 } else { legendi[['cex']] },
+						rot =  if (is.null(legendi[['labels.rot']])) { 0 } else { legendi[['labels.rot']] }
```
vertically align all this

pboutros commented on R/legend.grob.R:
```diff
@@ -80,154 +80,163 @@ legend.grob <- function(
+				color.key.grob$framevp$valid.just <- c(0,0.5);
+				
+				# need padding to line up with key style legends
+				color.key.grob$framevp$x = unit(0,'npc') + unit(1.69,'points');
```
why switching to using = for assignment?  And why 1.69?

pboutros commented on R/legend.grob.R:
```diff
@@ -80,154 +80,163 @@ legend.grob <- function(
+				color.key.grob$framevp$x = unit(0,'npc') + unit(1.69,'points');
+				
+				# set sizes
+				color.key.grob$framevp$height <- unit(legendi[['height']], 'lines');
+				color.key.grob$framevp$width <- unit(legendi[['width']], 'lines');
```
vertical alignment

pboutros commented on R/legend.grob.R:
```diff
@@ -80,154 +80,163 @@ legend.grob <- function(
+				color.key.grob$framevp$height <- unit(legendi[['height']], 'lines');
+				color.key.grob$framevp$width <- unit(legendi[['width']], 'lines');
+
+				if(!is.null(legendi[['pos.x']])) {
+					color.key.grob$framevp$x <- unit(legendi[['pos.x']],'npc');
+					}
```
probably convert into one-line so easier to read (and align with the one below for pos.y

pboutros commented on R/legend.grob.R:
```diff
@@ -80,154 +80,163 @@ legend.grob <- function(
+					row = 3 * (legend.row - 1) + 2,
+					col = 2 * (legend.col - 1) + 1,
+					height = max(unit(legendi[['height']], 'lines'),
+							legend.grob.final$framevp$layout$heights[3 * (legend.row - 1) + 1]),
```
why the change in indentation here?

pboutros commented on R/legend.grob.R:
```diff
@@ -80,154 +80,163 @@ legend.grob <- function(
+				        height = 1,
+				        border = legendi[['border']]
+				        );
+				} else if ('point' == typei) {
```
coding standard

pboutros commented on R/legend.grob.R:
```diff
@@ -80,154 +80,163 @@ legend.grob <- function(
+					row = 3 * (legend.row - 1) + 2,
+					col = 2 * (legend.col - 1) + 1
+					);
+				}
```
add examples of the new functionality in use

pboutros reviewed (changes requested):
Also make sure Caden and somebody with a lot of plotting experience reviews the PR.  I assume this passes all tests, produces identical output files for all examples, etc.?

cbakkeby commented:
Hi Paul, I should have clarified, I only modified lines 206-222. The others show up in the diff because I removed the switch block and removed one level of indentation lines 83-228.

stefaneng commented on man/legend.grob.Rd:
```diff
@@ -118,9 +118,14 @@ legends2 <- list(
         colours = c('firebrick3', 'lightgrey'),
-        labels = c('A label', 'A longer label')
+        labels = c('A label', 'A longer label'),
+        # Set dot size
+        cex = 1.5,
+        # More row spacing needed with dots
+        padding.text = 3
```
Is this something that could be changed into a default? Doesn't seem like overlapping text/dots would be a good default setting.

stefaneng reviewed (changes requested):
The diff for this PR is a bit hard to read since many lines only have indentation/whitespace changes. The code looks fine to me other than making the default padding such that the text and dots don't overlap.

stefaneng reviewed (approved):
@dan-knight LGTM at this point
